backend/tournesol/views/contributor_recommendations.py

```diff
@@ -0,0 +1,97 @@
+"""
+Overrides the Polls API for recommendations specific to one user
+"""
+from django.db.models import F, Prefetch, Sum
+from django.shortcuts import get_object_or_404
+from rest_framework.permissions import IsAuthenticated
+
+from core.models import User
+from tournesol.serializers.contributor_recommendations import ContributorRecommendationsSerializer
+
+from ..models import ContributorRating, Entity, Poll
+from ..views import PollRecommendationsBaseAPIView
+
+
+class ContributorRecommendationsBaseView(PollRecommendationsBaseAPIView):
+    """
+    A base class used to factorize behaviours common to all contributor
+    recommendations views.
+    """
+
+    queryset = Entity.objects.none()
+    serializer_class = ContributorRecommendationsSerializer
+
+    def annotate_with_total_score(self, queryset, request, poll: Poll, user):
+        criteria_weight = self._build_criteria_weight_condition(
+            request, poll, when="contributorvideoratings__criteria_scores__criteria"
+        )
+
+        queryset = queryset.prefetch_related(
+            Prefetch(
+                "contributorvideoratings",
+                queryset=ContributorRating.objects.filter(poll=poll, user=user),
+            ),
+            "contributorvideoratings__criteria_scores",
+        )
+
+        return queryset.annotate(
+            total_score=Sum(
+                F("contributorvideoratings__criteria_scores__score") * criteria_weight,
+            ),
+        )
+
+    def filter_unsafe(self, queryset, filters):
+        show_unsafe = filters["unsafe"]
+        if not show_unsafe:
+            # Ignore RECOMMENDATIONS_MIN_CONTRIBUTORS, only filter on the
+            # total score
+            queryset = queryset.filter(total_score__gt=0)
+
+        return queryset.distinct()
```

Comment:
Calling `.distinct()` is not related to the "unsafe" filter, right?  
If that's still necessary, it should be moved to the definition of the base queryset in `get_queryset()`.

Reply:
That's true, I'm going to check if it's required.

Reply:
This `DISTINCT` doesn't seem to be required in my opinion. 

The generated SQL uses a `GROUP BY "tournesol_entity"."id"` clause which must not generate duplicated rows.

I've manually ran the query locally and on staging by adding and removing a `DISTINCT` clause, each time the number of returned rows is the same.

Here the generated SQL query without `DISTINCT`:

```sql
SELECT 
  "tournesol_entity"."id", 
  "tournesol_entity"."uid", 
  "tournesol_entity"."type", 
  "tournesol_entity"."tournesol_score", 
  "tournesol_entity"."rating_n_ratings", 
  "tournesol_entity"."rating_n_contributors", 
  SUM(
    (
      "tournesol_contributorratingcriteriascore"."score" * CASE WHEN "tournesol_contributorratingcriteriascore"."criteria" = 'largely_recommended' THEN 10 WHEN "tournesol_contributorratingcriteriascore"."criteria" = 'reliability' THEN 10 WHEN "tournesol_contributorratingcriteriascore"."criteria" = 'pedagogy' THEN 10 WHEN "tournesol_contributorratingcriteriascore"."criteria" = 'importance' THEN 10 WHEN "tournesol_contributorratingcriteriascore"."criteria" = 'layman_friendly' THEN 10 WHEN "tournesol_contributorratingcriteriascore"."criteria" = 'entertaining_relaxing' THEN 10 WHEN "tournesol_contributorratingcriteriascore"."criteria" = 'engaging' THEN 10 WHEN "tournesol_contributorratingcriteriascore"."criteria" = 'diversity_inclusion' THEN 10 WHEN "tournesol_contributorratingcriteriascore"."criteria" = 'better_habits' THEN 10 WHEN "tournesol_contributorratingcriteriascore"."criteria" = 'backfire_risk' THEN 10 ELSE 0 END
    )
  ) AS "total_score" 
FROM 
  "tournesol_entity" 
  LEFT OUTER JOIN "tournesol_contributorrating" ON (
    "tournesol_entity"."id" = "tournesol_contributorrating"."entity_id"
  ) 
  LEFT OUTER JOIN "tournesol_contributorratingcriteriascore" ON (
    "tournesol_contributorrating"."id" = "tournesol_contributorratingcriteriascore"."contributor_rating_id"
  ) 
WHERE 
  (
    "tournesol_contributorrating"."is_public" 
    AND "tournesol_contributorrating"."poll_id" = 1 
    AND "tournesol_contributorrating"."user_id" = 1
  ) 
GROUP BY 
  "tournesol_entity"."id" 
HAVING 
  SUM(
    (
      "tournesol_contributorratingcriteriascore"."score" * CASE WHEN (
        "tournesol_contributorratingcriteriascore"."criteria" = 'largely_recommended'
      ) THEN 10 WHEN (
        "tournesol_contributorratingcriteriascore"."criteria" = 'reliability'
      ) THEN 10 WHEN (
        "tournesol_contributorratingcriteriascore"."criteria" = 'pedagogy'
      ) THEN 10 WHEN (
        "tournesol_contributorratingcriteriascore"."criteria" = 'importance'
      ) THEN 10 WHEN (
        "tournesol_contributorratingcriteriascore"."criteria" = 'layman_friendly'
      ) THEN 10 WHEN (
        "tournesol_contributorratingcriteriascore"."criteria" = 'entertaining_relaxing'
      ) THEN 10 WHEN (
        "tournesol_contributorratingcriteriascore"."criteria" = 'engaging'
      ) THEN 10 WHEN (
        "tournesol_contributorratingcriteriascore"."criteria" = 'diversity_inclusion'
      ) THEN 10 WHEN (
        "tournesol_contributorratingcriteriascore"."criteria" = 'better_habits'
      ) THEN 10 WHEN (
        "tournesol_contributorratingcriteriascore"."criteria" = 'backfire_risk'
      ) THEN 10 ELSE 0 END
    )
  ) > 0.0 
ORDER BY 
  "total_score" DESC, 
  "tournesol_entity"."id" DESC;
```

Reply:
Fixed by https://github.com/tournesol-app/tournesol/pull/644/commits/bb9dbebaede6f6aedbc990bba9c3d71439cbde2d.